### DISCUSSION TITLE
Apply Singleton pattern to EuRxManager

### DISCUSSION THREAD

**Discussion by phj0407** [euphony-io/euphony#254] [Ideas]
# As - Is 
<br/>

While I apply Euphony library for sample app project, I find using one more object can cause malfunciton. I think applying singleton pattern to EuRxManager can prevent potential issue.
<br/>

# To-Do
<img src="https://user-images.githubusercontent.com/89901283/189919959-490fd2c6-1ea9-4bb5-8b8e-f7e3b4926f1b.png" width="600"/>
<br/>

Now, EuRxManager have 3 constructor. 
```java
public EuRxManager() {
		mOption = EuOption.builder()
				.modeWith(EuOption.ModeType.DEFAULT)
				.encodingWith(EuOption.CodingType.BASE16)
				.modulationWith(EuOption.ModulationType.FSK)
				.build();
	}

	public EuRxManager(RxEngineType engineType) {
		rxEngineType = engineType;
		if(rxEngineType == RxEngineType.EUPHONY_NATIVE_ENGINE)
			nativeCore = EuNativeConnector.getInstance();
		else
			nativeCore = null;
	}

	public EuRxManager(EuOption.ModeType mode) {
		mOption = EuOption.builder()
				.modeWith(mode)
				.build();
	}
```
<br/>

I think `public EuRxManager()`,` public EuRxManager(EuOption.ModeType mode) ` can be combined. And initializing` nativeCore` can be solved seperately by add new method.
<br/>
<br/>

**Comment by designe** [comment]
@phj0407 Great Idea! 👍 
Actually, There are 3 important things about `Singleton`

1. It is only 1 instance.
2. it is created once!
3. no argument with creation function. because above two.

Please show me your code! 👍

**Reply by phj0407** [reply]
Hello, while I apply 'Singleton' to code with your advice, I have question about constructor. I saw existing constructor initialized `mOption` and class has function for initialize mOption - `setOption (EuOption mOption)`. Also, nativeCore is created depending on rxEngineType.
<br/>
```java
// Existing code
public EuRxManager() {
		mOption = EuOption.builder()
				.modeWith(EuOption.ModeType.DEFAULT)
				.encodingWith(EuOption.CodingType.BASE16)
				.modulationWith(EuOption.ModulationType.FSK)
				.build();
	}

public EuRxManager(RxEngineType engineType) {
		rxEngineType = engineType;
		if(rxEngineType == RxEngineType.EUPHONY_NATIVE_ENGINE)
			nativeCore = EuNativeConnector.getInstance();
		else
			nativeCore = null;
	}

public void setOption(EuOption mOption) {
		this.mOption = mOption;
	}

public void setRxEngineType(RxEngineType type) {
		this.rxEngineType = type;
	}

```
<br/>

So I'm curious that if I apply 'Singleton' , 
- do I have initialize default mOption like below code? 
- Is it okay to change 'setRxEngineType' ?
<br/>

```java
private EuOption mOption =  EuOption.builder()
				.modeWith(EuOption.ModeType.DEFAULT)
				.encodingWith(EuOption.CodingType.BASE16)
				.modulationWith(EuOption.ModulationType.FSK)
				.build();

private static class InnerInstanceClazz{
		private static final EuRxManager instance = new EuRxManager();
	}

public static EuRxManager getInstance(){
		return InnerInstanceClazz.instance;
	}

public void setRxEngineType(RxEngineType type) {
		this.rxEngineType = type;
		if(this.rxEngineType == RxEngineType.EUPHONY_NATIVE_ENGINE)
			nativeCore = EuNativeConnector.getInstance();
		else
			nativeCore = null;
	}
```

**Reply by designe** [reply]
@phj0407 You're right :)
actually, `signleton` shouldn't have paramaters. 
because, initialized once and re-use a bunch of times.

We need to make a setter function for `RxEngineType` like you suggested.

Looks good to me :)
